### ISSUE TITLE
Render behavior in component classes

### ISSUE THREAD

**Issue by dvisockas** [ViewComponent/view_component#569]
## Feature request
Would it make sense to have template code in the component class definition file (similar to how JSX works)?

I just thought that it would be cool to have something like a `render` method that could hold the template code in the class, not a separate file. Something like:

```ruby
class SomeComponent < ViewComponent::Base
  def render
    <<~ERB
      <h1><%= post.title %></h1>
      <p><%= post.content %></p>
    ERB
  end
end
```

### Motivation
React uses this in JSX. Also this would reduce total component file count by 2. However if someone would like to render js, css, etc. this would need methods like `render_js` or something (?) (or wouldn't make sense anymore)

Another drawback is the slim/haml indentation, as they would have to be aligned with the left most column of the file, since squiggly heredoc strips whitespace from each line of the string.

**Comment by joelhawksley** [comment]
@dvisockas thank you for filing this issue!

The original implementation of ViewComponent actually used almost this _exact_ syntax! We moved to using separate templates for a couple of reasons, including:

- GitHub doesn't syntax-highlight ERB in HEREDOCs 😬 
- The Ruby team raised some concerns about HEREDOCs and Ruby 3 types.


That being said, I'd love to find a way to go beyond our support for [Inline Components](https://viewcomponent.org/#inline-component). I think that in a lot of cases, having a single file for a component instead of two makes for a cleaner, simpler design.

**Comment by dvisockas** [comment]
It would be possible to write the view part using rails tag helpers, but that would be cruel. 
What were the concerns regarding HEREDOCs?

**Comment by patchwork-her** [comment]
How about letting me know when you do something before you take it upon yourself to such idocracy?

**Comment by joelhawksley** [comment]
> What were the concerns regarding HEREDOCs?

@dvisockas besides those I already mentioned, perhaps @soutaro can elaborate the concerns re: Ruby 3 typing?

**Comment by joelhawksley** [comment]
Closed by #1556
